插件工程执行gradlew clean assemblePlugin提示失败

FAILURE: Build failed with an exception.

* What went wrong:
Failed to notify task execution listener.
> com/android/build/gradle/internal/scope/TaskOutputHolder$TaskOutputType

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.internal.event.ListenerNotificationException: Failed to notify task execution listener.
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:86)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:340)
        at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:327)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at com.sun.proxy.$Proxy93.afterExecute(Unknown Source)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:261)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:249)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:238)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:663)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:597)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.NoClassDefFoundError: com/android/build/gradle/internal/scope/TaskOutputHolder$TaskOutputType
        at com.didi.virtualapk.hooker.MergeManifestsHooker.afterTaskExecute(MergeManifestsHooker.groovy:62)
        at com.didi.virtualapk.hooker.MergeManifestsHooker.afterTaskExecute(MergeManifestsHooker.groovy)
        at com.didi.virtualapk.hooker.GradleTaskHooker$afterTaskExecute$0.call(Unknown Source)
        at com.didi.virtualapk.hooker.GradleTaskHooker$afterTaskExecute$0.call(Unknown Source)
        at com.didi.virtualapk.hooker.TaskHookerManager$VirtualApkTaskListener.afterExecute(TaskHookerManager.groovy:66)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:370)
        at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:352)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        ... 26 more

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/BasePlugin.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/BasePlugin.groovy
@@ -7,6 +7,7 @@ import com.android.build.gradle.internal.TaskManager
 import com.android.build.gradle.internal.api.ApplicationVariantImpl
 import com.android.build.gradle.internal.variant.VariantFactory
 import com.android.builder.core.VariantType
+import com.didi.virtualapk.os.Build
 import com.didi.virtualapk.tasks.AssemblePlugin
 import com.didi.virtualapk.utils.Log
 import com.didi.virtualapk.utils.Reflect
@@ -43,14 +44,8 @@ public abstract class BasePlugin implements Plugin<Project> {
     @Override
     public void apply(Project project) {
         this.project = project
-        project.ext.set(Constants.GRADLE_3_1_0, false)
 
-        try {
-            Class.forName('com.android.builder.core.VariantConfiguration')
-        } catch (Throwable e) {
-            // com.android.tools.build:gradle:3.1.0
-            project.ext.set(Constants.GRADLE_3_1_0, true)
-        }
+        Build.initGradleVersion(project)
 
         AppPlugin appPlugin = project.plugins.findPlugin(AppPlugin)
 
@@ -76,7 +71,7 @@ public abstract class BasePlugin implements Plugin<Project> {
 
         project.extensions.create('virtualApk', VAExtention)
 
-        if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
+        if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_1_X)) {
             TaskManager taskManager = Reflect.on(appPlugin).field('taskManager').get()
             taskFactory = taskManager.getTaskFactory()
         } else {
@@ -105,7 +100,7 @@ public abstract class BasePlugin implements Plugin<Project> {
                         }
                     }
 
-                    if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
+                    if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_1_X)) {
                         taskFactory.configure("assemblePlugin", action)
                     } else {
                         taskFactory.named("assemblePlugin", action)
@@ -141,7 +136,7 @@ public abstract class BasePlugin implements Plugin<Project> {
 
             appPlugin.variantManager.productFlavors.each {
                 String variantName
-                if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
+                if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_1_X)) {
                     variantName = Reflect.on('com.android.build.gradle.internal.core.VariantConfiguration')
                             .call('computeFullName', it.key, buildType, VariantType.DEFAULT, null)
                             .get()

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/Constants.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/Constants.groovy
@@ -1,5 +1,0 @@
-package com.didi.virtualapk
-
-public final class Constants {
-    public static final String GRADLE_3_1_0 = 'va.gradle.3.1.0'
-}

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/VAHostPlugin.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/VAHostPlugin.groovy
@@ -7,6 +7,7 @@ import com.android.build.gradle.internal.pipeline.TransformTask
 import com.android.build.gradle.internal.publishing.AndroidArtifacts
 import com.android.build.gradle.internal.transforms.ProGuardTransform
 import com.android.build.gradle.tasks.ProcessAndroidResources
+import com.didi.virtualapk.os.Build
 import com.didi.virtualapk.utils.FileUtil
 import com.didi.virtualapk.utils.Log
 import com.didi.virtualapk.utils.Reflect
@@ -34,19 +35,13 @@ public class VAHostPlugin implements Plugin<Project> {
     public void apply(Project project) {
 
         this.project = project
-        project.ext.set(Constants.GRADLE_3_1_0, false)
 
-        try {
-            Class.forName('com.android.builder.core.VariantConfiguration')
-        } catch (Throwable e) {
-            // com.android.tools.build:gradle:3.1.0
-            project.ext.set(Constants.GRADLE_3_1_0, true)
-        }
-        
+        Build.initGradleVersion(project)
+
         //The target project must be a android application module
         if (!project.plugins.hasPlugin('com.android.application')) {
             Log.e(TAG, "application required!")
-            return;
+            return
         }
 
         vaHostDir = new File(project.getBuildDir(), "VAHost")
@@ -97,7 +92,7 @@ public class VAHostPlugin implements Plugin<Project> {
                 List<String> deps = new ArrayList<String>()
                 Log.i TAG, "Used compileClasspath: ${applicationVariant.name}"
                 Set<ArtifactDependencyGraph.HashableResolvedArtifactResult> compileArtifacts
-                if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
+                if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_1_X)) {
                     ImmutableMap<String, String> buildMapping = Reflect.on('com.android.build.gradle.internal.ide.ModelBuilder')
                             .call('computeBuildMapping', project.gradle)
                             .get()
@@ -168,7 +163,7 @@ public class VAHostPlugin implements Plugin<Project> {
     /**
      * Keep the host app resource id same with last publish, in order to compatible with the published plugin
      */
-     def keepResourceIds(variant) {
+    def keepResourceIds(variant) {
 
 
         def VIRTUAL_APK_DIR = new File([project.rootDir, 'virtualapk'].join(File.separator))
@@ -202,7 +197,8 @@ public class VAHostPlugin implements Plugin<Project> {
     }
 
 
-    def collectResourceEntries(final File mergeXml, final String projectResDir, final Map typeEntries) {
+    def collectResourceEntries(
+            final File mergeXml, final String projectResDir, final Map typeEntries) {
 
         collectAarResourceEntries(null, projectResDir, mergeXml, typeEntries)
 
@@ -211,14 +207,15 @@ public class VAHostPlugin implements Plugin<Project> {
         project.configurations.compile.resolvedConfiguration.resolvedArtifacts.each {
             if (it.extension == 'aar') {
                 def moduleVersion = it.moduleVersion.id
-                def resPath = new File(aarDir,"${moduleVersion.group}/${moduleVersion.name}/${moduleVersion.version}/res")
+                def resPath = new File(aarDir, "${moduleVersion.group}/${moduleVersion.name}/${moduleVersion.version}/res")
                 collectAarResourceEntries(moduleVersion.version, resPath.path, mergeXml, typeEntries)
             }
         }
     }
 
 
-    def collectAarResourceEntries(String aarVersion, String resPath, File mergeXml, final Map typeEntries) {
+    def collectAarResourceEntries(String aarVersion, String resPath, File mergeXml,
+                                  final Map typeEntries) {
         final def merger = new XmlParser().parse(mergeXml)
         def filter = aarVersion == null ? {
             it.@config == 'main' || it.@config == 'release'
@@ -244,9 +241,9 @@ public class VAHostPlugin implements Plugin<Project> {
                             def name = it.@name
                             if (type.endsWith('-array')) {
                                 type = 'array'
-                            } else if (type == 'item'){
+                            } else if (type == 'item') {
                                 type = it.@type
-                            } else if (type == 'declare-styleable'){
+                            } else if (type == 'declare-styleable') {
                                 return
                             }
                             def entrySet = getEntriesSet(type, typeEntries)
@@ -280,20 +277,20 @@ public class VAHostPlugin implements Plugin<Project> {
                 }
                 if (type == 'style') {
                     if (styleNameMap.containsKey(values[2])) {
-                        pw.println "\t<public type=\"${type}\" name=\"${styleNameMap.get(values[2])}\" id=\"${values[3]}\" />"
+                        pw.println "\t<public type=\"${type}\" name=\"${styleNameMap.get(values[2])}\" id=\"${values[3]}\"/>"
                     }
                     return
                 }
                 //ID does not filter and remains redundant
                 if (type == 'id') {
-                    pw.println "\t<public type=\"${type}\" name=\"${values[2]}\" id=\"${values[3]}\" />"
+                    pw.println "\t<public type=\"${type}\" name=\"${values[2]}\" id=\"${values[3]}\"/>"
                     return
                 }
 
                 //Only keep resources' Id that are present in the current project
                 Set entries = hostResourceEntries[type]
                 if (entries != null && entries.contains(values[2])) {
-                    pw.println "\t<public type=\"${type}\" name=\"${values[2]}\" id=\"${values[3]}\" />"
+                    pw.println "\t<public type=\"${type}\" name=\"${values[2]}\" id=\"${values[3]}\"/>"
                 } else {
                     if (entries == null) {
                         if (type != lastSplitType) {
@@ -302,7 +299,7 @@ public class VAHostPlugin implements Plugin<Project> {
                         }
 
                     } else {
-                        if (type != 'attr'){
+                        if (type != 'attr') {
                             println ">>>> ${type} : ${values[2]} is deleted"
                         }
 
@@ -328,7 +325,7 @@ public class VAHostPlugin implements Plugin<Project> {
     }
 
 
-    def Set<String> getEntriesSet (final String type, final Map typeEntries) {
+    def Set<String> getEntriesSet(final String type, final Map typeEntries) {
         def entries = typeEntries[type]
         if (entries == null) {
             entries = [] as Set<String>

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/MergeAssetsHooker.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/MergeAssetsHooker.groovy
@@ -2,7 +2,6 @@ package com.didi.virtualapk.hooker
 
 import com.android.build.gradle.api.ApkVariant
 import com.android.build.gradle.tasks.MergeSourceSetFolders
-import com.android.ide.common.res2.AssetSet
 import com.didi.virtualapk.collector.dependence.AarDependenceInfo
 import com.didi.virtualapk.utils.Log
 import com.didi.virtualapk.utils.Reflect
@@ -42,31 +41,31 @@ class MergeAssetsHooker extends GradleTaskHooker<MergeSourceSetFolders> {
         }
 
         Reflect reflect = Reflect.on(task)
-        reflect.set('assetSetSupplier', new FixedSupplier(this, reflect.get('assetSetSupplier'), strippedAssetPaths))
+        reflect.set('assetSetSupplier', new FixedSupplier<>(this, reflect.get('assetSetSupplier'), strippedAssetPaths))
     }
 
     @Override
     void afterTaskExecute(MergeSourceSetFolders task) {
     }
     
-    static class FixedSupplier implements Supplier<List<AssetSet>> {
+    static class FixedSupplier<T> implements Supplier<List<T>> {
 
         MergeAssetsHooker hooker
-        Supplier<List<AssetSet>> origin
+        Supplier<List<T>> origin
         Set<String> strippedAssetPaths
         
-        FixedSupplier(MergeAssetsHooker hooker, Supplier<List<AssetSet>> origin, Set<String> strippedAssetPaths) {
+        FixedSupplier(MergeAssetsHooker hooker, Supplier<List<T>> origin, Set<String> strippedAssetPaths) {
             this.hooker = hooker
             this.origin = origin
             this.strippedAssetPaths = strippedAssetPaths
         }
         
         @Override
-        List<AssetSet> get() {
-            List<AssetSet> assetSets = origin.get()
-            assetSets.removeIf(new Predicate<AssetSet>() {
+        List<T> get() {
+            List<T> assetSets = origin.get()
+            assetSets.removeIf(new Predicate<T>() {
                 @Override
-                boolean test(AssetSet assetSet) {
+                boolean test(T assetSet) {
                     boolean ret = strippedAssetPaths.contains(assetSet.sourceFiles.get(0).path)
                     if (ret) {
                         Log.i 'MergeAssetsHooker', "Stripped asset of artifact: ${assetSet} -> ${assetSet.sourceFiles.get(0).path}"

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/MergeManifestsHooker.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/MergeManifestsHooker.groovy
@@ -1,10 +1,10 @@
 package com.didi.virtualapk.hooker
 
 import com.android.build.gradle.api.ApkVariant
-import com.android.build.gradle.internal.scope.TaskOutputHolder
 import com.android.build.gradle.tasks.MergeManifests
-import com.didi.virtualapk.Constants
+import com.didi.virtualapk.os.Build
 import com.didi.virtualapk.collector.dependence.DependenceInfo
+import com.didi.virtualapk.support.ScopeCompat
 import com.didi.virtualapk.utils.Log
 import com.didi.virtualapk.utils.Reflect
 import groovy.xml.QName
@@ -58,27 +58,28 @@ class MergeManifestsHooker extends GradleTaskHooker<MergeManifests> {
      */
     @Override
     void afterTaskExecute(MergeManifests task) {
-        if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
-            File outputFile = Reflect.on('com.android.build.gradle.internal.scope.ExistingBuildElements')
-                    .call('from', TaskOutputHolder.TaskOutputType.MERGED_MANIFESTS, scope.getOutput(TaskOutputHolder.TaskOutputType.MERGED_MANIFESTS))
-                    .call('element', variantData.outputScope.mainSplit)
+        def MERGED_MANIFESTS = ScopeCompat.getArtifact(project, "MERGED_MANIFESTS")
+        if (Build.isSupportVersion(project,Build.VERSION_CODE.V3_1_X)) {
+            File outputFile = (Reflect.on('com.android.build.gradle.internal.scope.ExistingBuildElements')
+                    .call('from', MERGED_MANIFESTS, ScopeCompat.getArtifactFile(scope, project, MERGED_MANIFESTS))
+                    .call('element', variantData.outputScope.mainSplit))
                     .call('getOutputFile')
                     .get()
             rewrite(outputFile)
         } else {
-            variantData.outputScope.getOutputs(TaskOutputHolder.TaskOutputType.MERGED_MANIFESTS).each {
+            variantData.outputScope.getOutputs(MERGED_MANIFESTS).each {
                 rewrite(it.outputFile)
             }
         }
     }
-    
+
     void rewrite(File xml) {
         if (xml?.exists()) {
             final Node manifest = new XmlParser().parse(xml)
 
 
             manifest.application.each { application ->
-                [ 'icon', 'label', 'allowBackup', 'supportsRtl' ].each {
+                ['icon', 'label', 'allowBackup', 'supportsRtl'].each {
                     application.attributes().remove(new QName(MergeManifestsHooker.ANDROID_NAMESPACE, it))
                 }
             }
@@ -88,13 +89,13 @@ class MergeManifestsHooker extends GradleTaskHooker<MergeManifests> {
             })
         }
     }
-    
+
     private static class FixedArtifactCollection implements ArtifactCollection {
 
         private MergeManifestsHooker hooker
         private ArtifactCollection origin
         def stripAarNames
-        
+
         FixedArtifactCollection(MergeManifestsHooker hooker, ArtifactCollection origin, stripAarNames) {
             this.hooker = hooker
             this.origin = origin

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/PrepareDependenciesHooker.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/PrepareDependenciesHooker.groovy
@@ -5,7 +5,7 @@ import com.android.build.gradle.internal.ide.ArtifactDependencyGraph
 import com.android.build.gradle.internal.tasks.AppPreBuildTask
 import com.android.builder.model.Dependencies
 import com.android.builder.model.SyncIssue
-import com.didi.virtualapk.Constants
+import com.didi.virtualapk.os.Build
 import com.didi.virtualapk.collector.dependence.AarDependenceInfo
 import com.didi.virtualapk.collector.dependence.DependenceInfo
 import com.didi.virtualapk.collector.dependence.JarDependenceInfo
@@ -70,7 +70,7 @@ class PrepareDependenciesHooker extends GradleTaskHooker<AppPreBuildTask> {
             }
         }
         Dependencies dependencies
-        if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
+        if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_1_X)) {
             ImmutableMap<String, String> buildMapping = Reflect.on('com.android.build.gradle.internal.ide.ModelBuilder')
                     .call('computeBuildMapping', project.gradle)
                     .get()

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/ProcessResourcesHooker.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/hooker/ProcessResourcesHooker.groovy
@@ -3,14 +3,15 @@ package com.didi.virtualapk.hooker
 import com.android.build.gradle.AndroidConfig
 import com.android.build.gradle.AppExtension
 import com.android.build.gradle.api.ApkVariant
-import com.android.build.gradle.internal.scope.TaskOutputHolder
 import com.android.build.gradle.tasks.ProcessAndroidResources
 import com.android.sdklib.BuildToolInfo
-import com.didi.virtualapk.Constants
+import com.didi.virtualapk.os.Build
 import com.didi.virtualapk.aapt.Aapt
 import com.didi.virtualapk.collector.ResourceCollector
 import com.didi.virtualapk.collector.res.ResourceEntry
 import com.didi.virtualapk.collector.res.StyleableEntry
+import com.didi.virtualapk.support.ScopeCompat
+
 import com.didi.virtualapk.utils.FileUtil
 import com.didi.virtualapk.utils.Log
 import com.didi.virtualapk.utils.Reflect
@@ -59,15 +60,17 @@ class ProcessResourcesHooker extends GradleTaskHooker<ProcessAndroidResources> {
      */
     @Override
     void afterTaskExecute(ProcessAndroidResources par) {
-        if (project.extensions.extraProperties.get(Constants.GRADLE_3_1_0)) {
+
+        def PROCESSED_RES = ScopeCompat.getArtifact(project, "PROCESSED_RES")
+        if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_1_X)) {
             File outputFile = Reflect.on('com.android.build.gradle.internal.scope.ExistingBuildElements')
-                    .call('from', TaskOutputHolder.TaskOutputType.PROCESSED_RES, scope.getOutput(TaskOutputHolder.TaskOutputType.PROCESSED_RES))
+                    .call('from', PROCESSED_RES, ScopeCompat.getArtifactFile(scope, project, PROCESSED_RES))
                     .call('element', variantData.outputScope.mainSplit)
                     .call('getOutputFile')
                     .get()
             repackage(par, outputFile)
         } else {
-            variantData.outputScope.getOutputs(TaskOutputHolder.TaskOutputType.PROCESSED_RES).each {
+            variantData.outputScope.getOutputs(PROCESSED_RES).each {
                 repackage(par, it.outputFile)
             }
         }
@@ -198,18 +201,18 @@ class ProcessResourcesHooker extends GradleTaskHooker<ProcessAndroidResources> {
 
         pluginResources.keySet().each { resType ->
             def firstEntry = pluginResources.get(resType).get(0)
-            def typeEntry = [ type: "int", name: resType,
-                              id: parseTypeIdFromResId(firstEntry.resourceId),
-                              _id: parseTypeIdFromResId(firstEntry.newResourceId),
-                              entries: []]
+            def typeEntry = [type   : "int", name: resType,
+                             id     : parseTypeIdFromResId(firstEntry.resourceId),
+                             _id    : parseTypeIdFromResId(firstEntry.newResourceId),
+                             entries: []]
 
             pluginResources.get(resType).each { resEntry ->
                 typeEntry.entries.add([
-                        name : resEntry.resourceName,
-                        id : parseEntryIdFromResId(resEntry.resourceId),
-                        _id: parseEntryIdFromResId(resEntry.newResourceId),
-                        v : resEntry.resourceId, _v : resEntry.newResourceId,
-                        vs: resEntry.hexResourceId, _vs : resEntry.hexNewResourceId])
+                        name: resEntry.resourceName,
+                        id  : parseEntryIdFromResId(resEntry.resourceId),
+                        _id : parseEntryIdFromResId(resEntry.newResourceId),
+                        v   : resEntry.resourceId, _v: resEntry.newResourceId,
+                        vs  : resEntry.hexResourceId, _vs: resEntry.hexNewResourceId])
             }
 
             retainedTypes.add(typeEntry)
@@ -218,7 +221,7 @@ class ProcessResourcesHooker extends GradleTaskHooker<ProcessAndroidResources> {
         retainedTypes.sort { t1, t2 ->
             t1._id - t2._id
         }
-        
+
         return retainedTypes
     }
 
@@ -229,10 +232,10 @@ class ProcessResourcesHooker extends GradleTaskHooker<ProcessAndroidResources> {
     def convertStyleablesForAapt(List<StyleableEntry> pluginStyleables) {
         def retainedStyleables = []
         pluginStyleables.each { styleableEntry ->
-            retainedStyleables.add([vtype : styleableEntry.valueType,
+            retainedStyleables.add([vtype: styleableEntry.valueType,
                                     type : 'styleable',
-                                    key : styleableEntry.name,
-                                    idStr : styleableEntry.value])
+                                    key  : styleableEntry.name,
+                                    idStr: styleableEntry.value])
         }
         return retainedStyleables
     }

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/os/Build.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/os/Build.groovy
@@ -1,0 +1,44 @@
+package com.didi.virtualapk.os
+
+import org.gradle.api.Project
+
+final class Build {
+
+    public static final String GRADLE_VERSION_SDK_INT = 'va.gradle.version.sdk_int'
+
+    static interface VERSION_CODE {
+        int NONE = 0
+        int V3_1_X = 310
+        int V3_2_X = 320
+    }
+
+    static void initGradleVersion(Project project) {
+        if (findClass("com.android.build.gradle.internal.scope.InternalArtifactType")) {
+            project.ext.set(GRADLE_VERSION_SDK_INT, VERSION_CODE.V3_2_X)
+            return
+        }
+
+        if (!findClass("com.android.builder.core.VariantConfiguration")) {
+            project.ext.set(GRADLE_VERSION_SDK_INT, VERSION_CODE.V3_1_X)
+            return
+        }
+
+        project.ext.set(GRADLE_VERSION_SDK_INT, VERSION_CODE.NONE)
+    }
+
+
+    private static findClass(String className) {
+        try {
+            Class.forName(className)
+            return true
+        } catch (Throwable ignored) {
+            return false
+        }
+    }
+
+
+    static boolean isSupportVersion(Project project, int minVersion) {
+        return project.extensions.extraProperties.get(GRADLE_VERSION_SDK_INT) >= minVersion
+    }
+
+}

--- a/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/support/ScopeCompat.groovy
+++ b/virtualapk-gradle-plugin/src/main/groovy/com.didi.virtualapk/support/ScopeCompat.groovy
@@ -1,0 +1,30 @@
+package com.didi.virtualapk.support
+
+import com.android.build.gradle.internal.scope.VariantScope
+import com.didi.virtualapk.os.Build
+import com.didi.virtualapk.utils.Reflect
+import org.gradle.api.Project
+
+final class ScopeCompat {
+
+
+    def static getArtifact(Project project, String name) {
+        if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_2_X)) {
+            return Reflect.on("com.android.build.gradle.internal.scope.InternalArtifactType")
+                    .field(name).get()
+        } else {
+            return Reflect.on("com.android.build.gradle.internal.scope.TaskOutputHolder\$TaskOutputType")
+                    .field(name).get()
+        }
+    }
+
+
+    def static getArtifactFile(VariantScope scope, Project project, def artifact) {
+        if (Build.isSupportVersion(project, Build.VERSION_CODE.V3_2_X)) {
+            return scope.artifacts.getFinalArtifactFiles(artifact)
+        } else {
+            return scope.getOutput(artifact)
+        }
+    }
+
+}

